apps/server/src/shared/domain/rules/submission.rule.ts

```diff
@@ -0,0 +1,58 @@
+import { Injectable } from '@nestjs/common';
+import { Submission, User } from '../entity';
+import { IPermissionContext, PermissionTypes } from '../interface/permission';
+import { Actions } from './actions.enum';
+import { BasePermission } from './base-permission';
+import { TaskRule } from './task.rule';
+
+@Injectable()
+export class SubmissionRule extends BasePermission<Submission> {
+	constructor(private readonly taskRule: TaskRule) {
+		super();
+	}
+
+	public isApplicable(user: User, entity: PermissionTypes): boolean {
+		const isMatched = entity instanceof Submission;
+
+		return isMatched;
+	}
+
+	public hasPermission(user: User, entity: Submission, context: IPermissionContext): boolean {
+		const { action, requiredPermissions } = context;
+		const hasGeneralPermission = this.utils.hasAllPermissions(user, requiredPermissions);
+		const hasSpecificPermission = this.hasSpecificPermission(user, entity, action);
```

Comment:
In future we expect that read and write as scope actions are not enough. This is one of the bigger switches on this place.
For the developer and also for writing the test, refactorings and so on this switch should persist in code. 
It can be done behind the hasSpecificPermission or on this place, but it should not be hide in the code in lines 31 to 37.

Reply:
I've split up the logic for the different actions.